Mathlib/FieldTheory/IsPerfectClosure.lean

```diff
@@ -0,0 +1,421 @@
+
+section lift
+
+variable [PerfectRing M p] (H : ∀ x : L, ∃ (n : ℕ) (y : K), i y = x ^ p ^ n)
+
+theorem PerfectRing.lift_aux (x : L) : ∃ y : ℕ × K, i y.2 = x ^ p ^ y.1 := by
+  obtain ⟨n, y, h⟩ := H x
+  exact ⟨(n, y), h⟩
+
+/-- If `i : K →+* L` and `j : K →+* M` are ring homomorphisms of characteristic `p` rings, such that
+`i` is "purely inseparable" and `M` is a perfect ring, then one can define a map
+`L → M` which maps an element `x` of `L` to `y ^ (p ^ -n)` if `x ^ (p ^ n)` is equal to some
+element `y` of `K`. -/
+def PerfectRing.liftAux (x : L) : M :=
+  (frobeniusEquiv M p).symm^[(Classical.choose (lift_aux i p H x)).1]
+    (j (Classical.choose (lift_aux i p H x)).2)
+
+theorem PerfectRing.liftAux_self_apply [PerfectRing L p] (x : L) : liftAux i i p H x = x := by
+  rw [liftAux, Classical.choose_spec (lift_aux i p H x), ← iterate_frobenius,
+    iterate_frobeniusEquiv_symm_apply_iterate_frobenius]
+
+theorem PerfectRing.liftAux_self [PerfectRing L p] : liftAux i i p H = id :=
+  funext (liftAux_self_apply i p H)
+
+end lift
+
+end CommSemiring
+
+section CommRing
+
+variable [CommRing K] [CommRing L] [CommRing M]
+  (i : K →+* L) (j : K →+* M) (p : ℕ) [ExpChar K p] [ExpChar L p] [ExpChar M p]
+
+section lift
+
+variable [PerfectRing M p] (H : ∀ x : L, ∃ (n : ℕ) (y : K), i y = x ^ p ^ n)
+  (hk : RingHom.ker i ≤ pNilradical K p)
+
+/-- If `i : K →+* L` and `j : K →+* M` are ring homomorphisms of characteristic `p` rings, such that
+`i` is "purely inseparable" and whose kernel is contained in the `p`-nilradical of `K`, and `M` is a
+perfect ring, then `PerfectRing.liftAux` is well-defined. -/
+theorem PerfectRing.liftAux_apply (x : L) (n : ℕ) (y : K) (h : i y = x ^ p ^ n) :
+    liftAux i j p H x = (frobeniusEquiv M p).symm^[n] (j y) := by
+  rw [liftAux]
+  have h' := Classical.choose_spec (lift_aux i p H x)
+  set n' := (Classical.choose (lift_aux i p H x)).1
+  replace h := congr($(h.symm) ^ p ^ n')
+  rw [← pow_mul, mul_comm, pow_mul, ← h', ← map_pow, ← map_pow, ← sub_eq_zero, ← map_sub,
+    ← RingHom.mem_ker] at h
+  obtain ⟨m, h⟩ := mem_pNilradical.1 (hk h)
+  apply_fun _ using (injective_frobenius M p).iterate (m + n + n')
+  conv_lhs => rw [Function.iterate_add_apply, iterate_frobenius_apply_iterate_frobeniusEquiv_symm]
+  rw [add_assoc, add_comm n n', ← add_assoc,
+    Function.iterate_add_apply _ (m + n') n, iterate_frobenius_apply_iterate_frobeniusEquiv_symm,
+    iterate_frobenius, iterate_frobenius, ← sub_eq_zero, ← map_pow, ← map_pow, ← map_sub,
+    add_comm m, add_comm m, pow_add, pow_mul, pow_add, pow_mul, ← sub_pow_expChar_pow, h, map_zero]
+
+/-- If `i : K →+* L` and `j : K →+* M` are ring homomorphisms of characteristic `p` rings, such that
+`i` is "purely inseparable" and whose kernel is contained in the `p`-nilradical of `K`, and `M` is a
+perfect ring, then `PerfectRing.liftAux` is a ring homomorphism. -/
+def PerfectRing.lift : L →+* M where
+  toFun := liftAux i j p H
```

Comment:
Even though we could redefine [PerfectClosure.lift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PerfectClosure.html#PerfectClosure.lift) using this but it won't be as nice definitionally as it is now. The noncomputability of the current PerfectClosure.lift is all contained within `frobeniusEquiv`, while here `liftAux` uses choice to pick an `n`. If we change
```
pow_mem' : ∀ x : L, ∃ (n : ℕ) (y : K), i y = x ^ p ^ n
```
to 
```
pow_mem' : (x : L) → trunc {ny : ℕ × K // i ny.2 = x ^ p ^ ny.1}
```
then we can make PerfectRing.lift definitionally as nice as PerfectClosure.lift, but I think it's overly complicated for little gain, so I'd suggest just leave PerfectClosure.lift as is.

Reply:
This `PerfectRing.lift` is not the same as the `PerfectClosure.lift`. I'm not defining `IsPerfectClosure.lift` yet, will do it later.

Reply:
I think that the original `PerfectClosure.lift` is a special case of `IsPRadical.lift`, here `IsPRadical` is a generalization of `IsPurelyInseparable` to rings.

Reply:
Ohhhh I think you're right. I added `IsPRadical` and `PerfectRing.liftEquiv` which is indeed a generalization of `PerfectClosure.lift`.